### PR TITLE
Hint/enhancement/85 notification refactor

### PR DESCRIPTION
**Pull request by dinjas**
Addresses #85. 

This includes a pretty major refactoring of the notification system. Some nice side effects are:

- it won't attempt to send notifications to users without a phone number
- it won't add users without phone numbers to the `notified_user_ids` of a `Need`
- methods are much easier to test (and are tested)

### CODE REVIEW

**Comment by dinjas** [comment]
Quick update here: still a work in progress, but making progress. I am going to be out of town for a few days, but will pick this back up when I return. I have work left to do on specs, and need to double check some logic against the master branch regarding which users should be notified for certain events.

**Comment by benjaminwood** [comment]
I noticed that the related error has occurred nearly 8k times now: https://app.honeybadger.io/projects/59936/faults/48746325#notice-summary

@dinjas what remains on this PR?

As a stopgap, maybe we should put the guard clause back in the worker which returns if the phone number is not present/valid (which I had previously removed). Not meaning to rush you, just feel we should do *something* if the refactor is still a ways out.

**Comment by dinjas** [comment]
I think putting the guard clause in might not be a bad idea. Feels like I probably have close to a day left to wrap this up - mostly updating specs and addressing any fallout.

**Comment by nvick** [comment]
We also should limit retries for message sending.

**Comment by dinjas** [comment]
🎉

File: app/lib/services/notifications/needs.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# require 'app/lib/services/notifications/needs/message.rb'
+# require 'app/lib/services/notifications/needs/recipients.rb'
+
+module Services
+  module Notifications
+    class Needs
+      # include Adamantium::Flat
```
Comment (dinjas) [inline app/lib/services/notifications/needs.rb]: remove

File: app/lib/services/notifications/needs/recipients.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+          klass = case action
+                    when :create
+                      Create
+                    when :update
+                      Update
+                    when :destroy
+                      Destroy
+                  end
+
+          klass.new(need).recipients.select(&:notifiable?)
```
Comment (dinjas) [inline app/lib/services/notifications/needs/recipients.rb]: see if tests pass without this

Comment (dinjas) [reply app/lib/services/notifications/needs/recipients.rb]: lol
without `notifiable?`

**Pull request merged**